maven-core/src/main/java/org/apache/maven/classrealm/DefaultClassRealmManager.java

```diff
@@ -152,14 +150,15 @@ private ClassRealm createRealm(
             List<String> parentImports,
             Map<String, ClassLoader> foreignImports,
             List<Artifact> artifacts) {
-        Set<String> artifactIds = new LinkedHashSet<>();
+        Set<String> artifactIds = new LinkedHashSet<>(artifacts == null ? 0 : artifacts.size());
+        List<ClassRealmConstituent> constituents = new ArrayList<>(artifacts == null ? 0 : artifacts.size());
 
-        List<ClassRealmConstituent> constituents = new ArrayList<>();
-
-        if (artifacts != null) {
+        if (artifacts != null && !artifacts.isEmpty()) {
             for (Artifact artifact : artifacts) {
                 if (!isProvidedArtifact(artifact)) {
-                    artifactIds.add(getId(artifact));
+                    if (logger.isDebugEnabled()) {
+                        artifactIds.add(getId(artifact));
+                    }
```

Comment:
I would not make the computation of the artifactIds condition to the log level.

Reply:
Let me try to explain what was the point of this change. The artifactIds list is computed only for debugging purpose. By adding the condition on the log level it is avoiding the computation of building this list if logging is not going to happen. The same for includedIds in your other comment. They are both used only to compute the debug logs.

One other comment, while trying to reduce the complexity of this code and limit the number of isDebugEnabled()  conditions, I came to the conclusion that something in this logic is fishy. If you look closely on the createRealm() method where the artifactIds list is populated and in the populateRealm() where the includedIds is populated. The only elements that would be in the artifactIds list and not in the includedIds list are artifacts with getFile() null value. If I am right the logic of this code can be reduced a lot, both limiting the isDebugEnabled()  conditions and removing some list constructions in the populateRealm() method.

Please tell me what do you think about this idea. Did I miss something in the code? I can patch the PR with the simplified code if you agree.

Reply:
I made a commit with the proposed changes (direct link [here](maven-core/src/main/java/org/apache/maven/classrealm/DefaultClassRealmManager.java)), please comment. This is reducing the number of isDebugEnabled() blocks and the collections used only for logging informations. The business logic is unchanged and the logged information is the same.

Reply:
LGTM